src/launchpad/kafka.py

```diff
@@ -67,14 +68,89 @@ def __init__(
+
         # Override SIGCHLD handler - child exits are expected with maxtasksperchild=1
         signal.signal(
             signal.SIGCHLD,
             lambda signum, frame: logger.debug(f"Worker process exited normally (SIGCHLD {signum})"),
         )
 
+    def poll(self) -> None:
+        if self._batch_timeout is not None:
+            try:
+                self._check_batch_timeouts()
+            except Exception as e:
+                logger.error(f"Error checking batch timeouts: {e}", exc_info=True)
+
+        super().poll()
+
+    def _check_batch_timeouts(self) -> None:
+        """
+        Check if any in-flight batches have exceeded timeout and terminate those workers.
+
+        This accesses parent class private members via name mangling. If parent class
+        changes these member names, this will break.
+        """
+
+        try:
+            processes = self._RunTaskWithMultiprocessing__processes
+            pool = self._RunTaskWithMultiprocessing__pool
+            invalid_messages = self._RunTaskWithMultiprocessing__invalid_messages
+        except AttributeError:
+            logger.exception("Failed to access parent class private members - please check if the Arroyo API changed:")
+            return
+
+        # Track batch times to know which batch is timed out
+        # Make sure to clean up finished/timed out batches so this doesn't grow unbounded
+        if processes:
+            current_batch_ids = {id(batch) for batch in processes}
+            completed_batch_ids = [bid for bid in self._batch_submit_times if bid not in current_batch_ids]
+            for batch_id in completed_batch_ids:
+                del self._batch_submit_times[batch_id]
+        else:
+            self._batch_submit_times.clear()
+            return
+
+        # Check the first batch, we only support batchsize=1 right now
+        first_batch = processes[0]
+        batch_id = id(first_batch)
+
+        if batch_id not in self._batch_submit_times:
+            self._batch_submit_times[batch_id] = time.time()
+            return
+
+        elapsed = time.time() - self._batch_submit_times[batch_id]
+
+        if elapsed > self._batch_timeout:
+            logger.error(f"Batch exceeded timeout of {self._batch_timeout}s (elapsed={elapsed:.2f}s).")
+
+            input_batch = first_batch[0]
+
+            pool.close()  # Terminates all workers
+            pool.maybe_create_pool()  # Recreate fresh pool
+
+            # Remove timed-out batch from queue
+            processes.popleft()
+
+            # Convert batch messages to InvalidMessages for DLQ
+            for idx, message in input_batch:
+                invalid_msg = InvalidMessage(
+                    message.value.partition,
+                    message.value.offset,
+                    TimeoutError(f"Batch processing exceeded {self._batch_timeout}s timeout"),
+                )
+                invalid_messages.append(invalid_msg)
```

Comment:
This part I wasn't entirely sure about, but I didn't think I could raise `InvalidMessage` from this part of the code.